fix(Core/Creature): Creatures will periodically call for assistance.

Chero opened the pull request:
## Changes Proposed:
Chasing and fleeing creatures will periodically call for assistance.
Humanoids, Demons, Undeads, and certain social beasts (wolf, cat, spider, crocolisk) are affected.

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/4976 https://github.com/azerothcore/azerothcore-wotlk/issues/2144


## Tests Performed:
- Build & Implementation


## How to Test the Changes:
1) Fear a creature and watch how it picks up nearby friends
2) Pull a distant creature and watch how it picks up nearby friends during its path to you


## Target Branch(es):
- [x] Master


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

Hacki95 commented:
> Humanoids, Demons, Undeads, and certain social beasts (wolf, cat, spider, crocolisk) are affected.

Do you have some kind of source for this? I remember them calling for help, but was it really only special creature types and not every creature?

T1ti commented:
> > Humanoids, Demons, Undeads, and certain social beasts (wolf, cat, spider, crocolisk) are affected.
> 
> Do you have some kind of source for this? I remember them calling for help, but was it really only special creature types and not every creature?

Some creatures have special call for help abilities with really long range, but there is a global mechanic that creatures will always aggro other nearby creatures(which is missing right now and being added with this PR)

T1ti commented:
I'm really not sure if it's supposed to be race/creature specific like this PR implements it tho ? 
Imo it's just any creatures that have factions that can help each other, the same as when you pull a creature and it calls nearby creatures

Hacki95 commented:
> I'm really not sure if it's supposed to be race/creature specific like this PR implements it tho ?
> Imo it's just any creatures that have factions that can help each other, the same as when you pull a creature and it aggro(calls) nearby creatures
> Need some research to confirm that

This PR only implements it for certain creature types thats why I asked for a source. Basing it on faction would be another way which makes sense in my eyes but I am not sure.

T1ti commented:
It should just work the same way it works for the initial aggro

It's simple, when a creature is in combat it permanantly ticks for help around him
I think the initial aggro tick should even be using the same as the initial combat tick, they shouldn't be separate I think

Hacki95 commented:
I added the needs feedback label as I think we need a little bit more research into this.

T1ti commented:
Gonna test with "non social" creatures on classic to confirm dragon etc also have that call for help.

I talked with brotalnia and figured out most of that mechanic, gonna make a detailed post in a bit

T1ti commented:
I can confirm all creatures call for help on Classic, it's just faction based so that "social creature" system is not needed
![image](https://user-images.githubusercontent.com/40864460/113061085-d3961780-91b1-11eb-9161-39882ca553f0.png)

So I talked a lot with brotalnia about the call for helps (he researched and implemented it on vmangos) and also did some testing on Classic

-The initial aggro call for help on pull and the ticking call for help are different mechanics.
-The first call for help will link creatures leashes but the ticking doesn't.
-They also have different timers, the initial call for help has a 2sec timer, the ticking has a 3sec delay(which is why creatures pulled from the first call for help don't immediatly aggro around them too)

Video from Classic I just recorded where u can see a first creature doing the initial call for help on a caster, and that caster then does a ticking call for help after a 3sec delay on the 3rd creature that was only in range of the 2nd(hope you're following lol)
Here are the timers I found frame by frame(sadly only 30fps, 60fps would have been more accurate)
https://i.gyazo.com/bd560f0be3b00fd3bdf2c33412097a7f.mp4
**1st aggro 17.499
initial call for help aggro19.605 (2sec)
chain pul(tick from 2nd creature)l 22.808**

So roughly 2sec delay for the initial call for help, 3sec for the tick

pklloveyou commented:
![image](https://user-images.githubusercontent.com/18535853/113071380-3231b480-91f7-11eb-9502-781eeebeec05.png)
Is that right?

pklloveyou commented:
I think we should add a Los check
![image](https://user-images.githubusercontent.com/18535853/113071601-ab310c00-91f7-11eb-9fe1-7ee416142e4b.png)

pklloveyou commented:
![image](https://user-images.githubusercontent.com/18535853/113071960-6c4f8600-91f8-11eb-8792-dbf06c6e6a6f.png)

T1ti commented:
sorry dude we don't understand what those paints mean, they're even worse then mine

T1ti commented:
```
 // Must be chasing or fleeing target
    if (!HasUnitState(UNIT_STATE_FLEEING | UNIT_STATE_CHASE))
        return false;
```

I don't think this is correct, because chasing means the creature is out of range and running

For a static creature like the skeleton mage in that video, he is neither chasing or fleeing but should still tick.
It's basically just "in combat"

pklloveyou commented:
![image](https://user-images.githubusercontent.com/18535853/113072943-bb96b600-91fa-11eb-9452-6280ea254fae.png)

pklloveyou commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -3152,3 +3166,45 @@ void Creature::SetCannotReachTarget(bool cannotReach)
+    if (HasUnitState(UNIT_STATE_STUNNED | UNIT_STATE_DIED | UNIT_STATE_CONFUSED | UNIT_STATE_POSSESSED))
+        return false;
+
+    // Must be chasing or fleeing target
+    if (!HasUnitState(UNIT_STATE_FLEEING | UNIT_STATE_CHASE))
+        return false;
```
This can happen in a multiplayer operation, I think we should add another variant and vertigo

T1ti commented:
I guess you mean that those are missing checks he should add ? they look good

pklloveyou commented:
![image](https://user-images.githubusercontent.com/18535853/113073315-732bc800-91fb-11eb-8d6e-9e181623749d.png)

pklloveyou commented:
![image](https://user-images.githubusercontent.com/18535853/113073379-8f2f6980-91fb-11eb-9f98-478bbc706798.png)
I don't understand why it's done here

T1ti commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -3152,3 +3166,45 @@ void Creature::SetCannotReachTarget(bool cannotReach)
+bool Creature::CanPeriodicallyCallForAssistance() const
+{
+    // Unable to call for assistance
+    if (HasUnitState(UNIT_STATE_STUNNED | UNIT_STATE_DIED | UNIT_STATE_CONFUSED | UNIT_STATE_POSSESSED))
```
I think creatures should still aggro while stunned/confused, needs testing I guess

pklloveyou replied:
ok

pklloveyou replied:
Do you know the status code of "change sheep"?

T1ti replied:
I'm pretty sure sheeps aggro, the logic is more like the other creature notices that this creature is attacked/fightning rather than the creature itself calling the others

T1ti replied:
Just asked a few friends that play mage, they confirm that sheeps aggro

pklloveyou replied:
I mean, creatures should not call stun and sheep changing creatures to fight

pklloveyou commented:
```c++
bool Creature::CanPeriodicallyCallForAssistance(Creature* creature) const
{
    if (!IsAlive())
        return false;

    if (IsInCombat())
        return false;

    //Unable to call for assistance
    if (HasUnitState(UNIT_STATE_STUNNED | UNIT_STATE_DIED | UNIT_STATE_CONFUSED | UNIT_STATE_POSSESSED))
        return false;

    //Must be chasing or fleeing target
    if (!HasUnitState(UNIT_STATE_FLEEING | UNIT_STATE_CHASE))
        return false;


    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_DISABLE_MOVE
        | UNIT_FLAG_NON_ATTACKABLE
        | UNIT_FLAG_NOT_SELECTABLE
        | UNIT_FLAG_CONFUSED))
        return false;

    if (HasUnitState(UNIT_STATE_EVADE
        | UNIT_STATE_ROOT
        | UNIT_STATE_DISTRACTED
        | UNIT_STATE_CONFUSED_MOVE))
        return false;

    if (HasReactState(REACT_PASSIVE))
        return false;

    if (!InSamePhase(creature))
        return false;

    if (!IsWithinLOSInMap(creature))
        return false;

    if (!CanHaveThreatList())
        return false;

    //Smart creatures
    switch (GetCreatureTemplate()->type)
    {
    case CREATURE_TYPE_HUMANOID:
    case CREATURE_TYPE_UNDEAD:
    case CREATURE_TYPE_DEMON:
        return true;
    default:
        break;
    }

    //Social type beasts
    if (GetCreatureTemplate()->type == CREATURE_TYPE_BEAST)
    {
        switch (GetCreatureTemplate()->family)
        {
        case CREATURE_FAMILY_WOLF:
        case CREATURE_FAMILY_CAT:
        case CREATURE_FAMILY_CROCOLISK:
        case CREATURE_FAMILY_SPIDER:
            return true;
        default:
            break;
        }
    }
    return false;
}
```
Can we consider this?

Chero commented:
Thank you guys for the feedback and review.

I have updated my PR based on that. Also, I have updated the default assistance delay in the config options and added a new config option for the periodic one if one deems it necessary to disable this.

pklloveyou commented on src/server/game/Entities/Creature/Creature.h:
```diff
@@ -813,6 +815,8 @@ class Creature : public Unit, public GridObject<Creature>, public MovableMapObje
 private:
     void ForcedDespawn(uint32 timeMSToDespawn = 0);
 
+    bool CanPeriodicallyCallForAssistance() const;
```
[[nodiscard]] bool CanPeriodicallyCallForAssistance() const;

Chero replied:
Resolved.

Hacki95 commented:
> 
> 
> Thank you guys for the feedback and review.
> 
> I have updated my PR based on that. Also, I have updated the default assistance delay in the config options and added a new config option for the periodic one if one deems it necessary to disable this.

Great work! Sorry for instantly coming with demands for your first PR here :p

Chero commented:
> > Thank you guys for the feedback and review.
> > I have updated my PR based on that. Also, I have updated the default assistance delay in the config options and added a new config option for the periodic one if one deems it necessary to disable this.
> 
> Great work! Sorry for instantly coming at you with a bunch of demands for your first PR here.
> 
> If you think it is ready to be reviewed let me know.

Pretty much, I have nothing to add at this point.

T1ti commented on src/server/game/World/World.cpp:
```diff
@@ -1036,7 +1036,8 @@ void World::LoadConfigSettings(bool reload)
     m_float_configs[CONFIG_CREATURE_FAMILY_FLEE_ASSISTANCE_RADIUS] = sConfigMgr->GetOption<float>("CreatureFamilyFleeAssistanceRadius", 30.0f);
     m_float_configs[CONFIG_CREATURE_FAMILY_ASSISTANCE_RADIUS]      = sConfigMgr->GetOption<float>("CreatureFamilyAssistanceRadius", 10.0f);
-    m_int_configs[CONFIG_CREATURE_FAMILY_ASSISTANCE_DELAY]         = sConfigMgr->GetOption<int32>("CreatureFamilyAssistanceDelay", 1500);
+    m_int_configs[CONFIG_CREATURE_FAMILY_ASSISTANCE_DELAY]         = sConfigMgr->GetOption<int32>("CreatureFamilyAssistanceDelay", 2000);
+    m_int_configs[CONFIG_CREATURE_FAMILY_ASSISTANCE_PERIOD]        = sConfigMgr->GetOption<int32>("CreatureFamilyAssistancePeriod", 3000);
```
I don't think that you understand what I meant when I posted the 2sec and 3sec timers
-2sec is the delay for the original call for help when you pull a creature, which is another different mechanic than this
-3sec is the delay before the creature starts ticking with this mechanic, I have no idea how often they tick though

Chero replied:
In this case, we need some values here that are good to go.

Helias commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -626,6 +626,22 @@ void Creature::Update(uint32 diff)
+                        }
+                        m_assistanceTimer = sWorld->getIntConfig(CONFIG_CREATURE_FAMILY_ASSISTANCE_PERIOD);
+                    }
+                    else
+                        m_assistanceTimer -= diff;
```
add brackets

FrancescoBorzi replied:
@Chero

pklloveyou commented:
good job

pklloveyou commented:
Tested

FrancescoBorzi commented:
what's the status of this? can it be merged?

T1ti commented:
need testers I guess

FrancescoBorzi commented:
@azerothcore/testers can anyone please look at this?

anguaive commented:
I did some testing on this. Both #4976 #2144 are fixed. Feared creatures dart around randomly, and similar creatures near them get mad at me. Pulling mobs with Taunt will also aggro similar creatures along the pulled mob's path. 

I don't think I was able to capture the chain pulling behavior (when the 2nd mob calls for help that is out of range of the first mob). That doesn't mean it doesn't work, it's just kind of hard to replicate that kind of a situation.

### A few clips:
(feared mob pulls 2 others)
https://user-images.githubusercontent.com/39011611/119352995-a1c99a80-bca2-11eb-890c-443d7963daf7.mp4

(watch the dude in front of the tent)
https://user-images.githubusercontent.com/39011611/119349290-4ac1c680-bc9e-11eb-93d3-aa5b623e2069.mp4

(same thing here)
https://user-images.githubusercontent.com/39011611/119350772-03d4d080-bca0-11eb-9c56-ef04ca8a7cce.mp4

(same here)
https://user-images.githubusercontent.com/39011611/119350869-1e0eae80-bca0-11eb-9fce-625dfe6ca636.mp4

(Taunt pull)
https://user-images.githubusercontent.com/39011611/119350824-12bb8300-bca0-11eb-910b-046b7f608875.mp4

FrancescoBorzi commented:
@anguaive thank you very much for testing this!

@T1ti @Hacki95 @pklloveyou is this ready to be merged or you still have some notes?

Hacki95 commented:
Nothing to add from me.

T1ti commented:
> I did some testing on this. Both #4976 #2144 are fixed. Feared creatures dart around randomly, and similar creatures near them get mad at me. Pulling mobs with Taunt will also aggro similar creatures along the pulled mob's path.
> 
> I don't think I was able to capture the chain pulling behavior (when the 2nd mob calls for help that is out of range of the first mob). That doesn't mean it doesn't work, it's just kind of hard to replicate that kind of a situation.
> 
> ### A few clips:
> (feared mob pulls 2 others)
> https://user-images.githubusercontent.com/39011611/119352995-a1c99a80-bca2-11eb-890c-443d7963daf7.mp4
> 
> (watch the dude in front of the tent)
> https://user-images.githubusercontent.com/39011611/119349290-4ac1c680-bc9e-11eb-93d3-aa5b623e2069.mp4
> 
> (same thing here)
> https://user-images.githubusercontent.com/39011611/119350772-03d4d080-bca0-11eb-9c56-ef04ca8a7cce.mp4
> 
> (same here)
> https://user-images.githubusercontent.com/39011611/119350869-1e0eae80-bca0-11eb-9fce-625dfe6ca636.mp4
> 
> (Taunt pull)
> https://user-images.githubusercontent.com/39011611/119350824-12bb8300-bca0-11eb-910b-046b7f608875.mp4

the aggro radius in the first video looks too large, otherwise it looks fine

FrancescoBorzi commented:
@T1ti the radius being too large is something changed by this PR or was it already like that?

T1ti commented:
I'm not seeing any radius in this PR so I expect it's using existing code for it

Sedatyf commented:
Nice work @anguaive on the test!
Nice work @Chero on this PR!

T1ti commented:
I watched the video again and realised I was being dumb, the first aggro in the first clip is just the regular on pull aggro, it's not caused by the fear
